ckan/migration/migrate_package_activity.py

```diff
@@ -0,0 +1,261 @@
+# args quickly.
+
+_context = None
+
+
+def get_context():
+    from ckan import model
+    import ckan.logic as logic
+    global _context
+    if not _context:
+        user = logic.get_action(u'get_site_user')(
+            {u'model': model, u'ignore_auth': True}, {})
+        _context = {u'model': model, u'session': model.Session,
+                    u'user': user[u'name']}
+    return _context
+
+
+def num_unmigrated(engine):
+    num_unmigrated = engine.execute('''
+        SELECT count(*) FROM activity a JOIN package p ON a.object_id=p.id
+        WHERE a.activity_type IN ('new package', 'changed package')
+        AND a.data NOT LIKE '%%{"actor"%%'
+        AND p.private = false;
+    ''').fetchone()[0]
+    return num_unmigrated
+
+
+def num_activities_migratable():
+    from ckan import model
+    num_activities = model.Session.execute(u'''
+    SELECT count(*) FROM activity a JOIN package p ON a.object_id=p.id
+    WHERE a.activity_type IN ('new package', 'changed package')
+    AND p.private = false;
+    ''').fetchall()[0][0]
+    return num_activities
+
+
+def migrate_all_datasets():
+    import ckan.logic as logic
+    dataset_names = logic.get_action(u'package_list')(get_context(), {})
+    num_datasets = len(dataset_names)
+    errors = defaultdict(int)
+    for i, dataset_name in enumerate(dataset_names):
+        print(u'\n{}/{} dataset: {}'.format(i + 1, num_datasets, dataset_name))
+        migrate_dataset(dataset_name, errors)
+    print(u'Migrated:')
+    print(u'  {} datasets'.format(len(dataset_names)))
+    num_activities = num_activities_migratable()
+    print(u'  with {} activities'.format(num_activities))
+    print_errors(errors)
+
+
+def migrate_dataset(dataset_name, errors):
+    # monkey patch the legacy versions of code back into CKAN - so it has the
+    # revision functionality needed for this migration
+    import ckan.lib.dictization.model_dictize as model_dictize
+    try:
+        import ckan.migration.revision_legacy_code as revision_legacy_code
+    except ImportError:
+        # convenient to look for it in the current directory if you just
+        # download these files because you are upgrading an older ckan
+        import revision_legacy_code
+    model_dictize.package_dictize = \
+        revision_legacy_code.package_dictize_with_revisions
+
+    import ckan.logic as logic
+    from ckan import model
+
+    context = get_context()
+    # 'hidden' activity is that by site_user, such as harvests, which are
+    # not shown in the activity stream because they can be too numerous.
+    # However these do have Activity objects, and if a hidden Activity is
+    # followed be a non-hidden one and you look at the changes of that
+    # non-hidden Activity, then it does a diff with the hidden one (rather than
+    # the most recent non-hidden one), so it is important to store the
+    # package_dict in hidden Activity objects.
```

Comment:
ok I was confused by hidden but that makes sense. How about calling these "system" activity or "automated" activity instead of "hidden"?

Reply:
I should have added that it's not just changes by the site_user, but you can configure other users' changes to hide using `ckan.hide_activity_from_users` config option: https://github.com/ckan/ckan/pull/1330/files It's a bit woolly, and "hidden" is not the best name, but I didn't want to stray from the existing language of this functionality, rooted in the config option. I think another option would be to call it "activity_by_filtered_out_users" which comes from the existing function `_activity_stream_get_filtered_users`, but it's a bit of a mouthful. Alternatively I rename all these things, bar the config option?